tmt/utils.py

```diff
@@ -7270,3 +7271,80 @@ def render_rst(text: str, logger: Logger) -> str:
     document.walkabout(visitor)
 
     return visitor.rendered
+
+
+def jira_link(
+        nodes: list[Union['tmt.base.Test', 'tmt.base.Plan', 'tmt.base.Story']],
+        links: 'tmt.base.Links',
+        separate: Optional[bool] = False) -> None:
+    """ Link the object to Jira issue and create the URL to tmt web service """
+
+    def create_url(tmt_object: 'tmt.base.Core') -> list[str]:
+        url: list[str] = []
+        # Get the fmf id of the object
+        if isinstance(tmt_object, tmt.base.Test):
+            fmfid = tmt_object.fmf_id
+            tmt_type = "test"
+        elif isinstance(tmt_object, tmt.base.Plan):
+            fmfid = tmt_object.fmf_id
+            tmt_type = "plan"
+        elif isinstance(tmt_object, tmt.base.Story):
+            # TODO: not supported by the service yet
+            fmfid = tmt_object.fmf_id
+            tmt_type = "story"
+        url.append(f'{tmt_type}-url={fmfid.url}')
+        url.append(f'{tmt_type}-name={fmfid.name}')
+        if fmfid.path is not None:
+            url.append(f'{tmt_type}-path={fmfid.path}')
+        if fmfid.ref is not None:
+            url.append(f'{tmt_type}-ref={fmfid.ref}')
+        return url
+
+    def construct_url_from_list(url: str, url_parts: list[str]) -> str:
+        joined_url = '&'.join(url_parts)
+        joined_url = url + joined_url
+        # Ask for HTML format (WIP, can be changed once FE web server is implemented)
+        joined_url += '&format=html'
+        return joined_url
```

Comment:
A couple of suggestions for URL generation: you spend too much time working out the low-level details, "manually" constructing the URL which is error-prone and easy to do incorrectly. I cannot recommend https://docs.python.org/3/library/urllib.parse.html#module-urllib.parse enough :)

1. `create_url()` is not creating URL, it's collecting its query components. The name is therefore misleading.
2. these components are pretty much key/value pairs, collect them as such. If the goal is to collect and construct URL query components, then use the best fitting data type for the goal, do not worry about the fact these components need to be encoded and merged into a string - such tasks are best performed at the boundary between your code and the outside world. The best-fitting data type is a simple dictionary:
   ```python
   tmt_type = tmt_object.__class__.__name__.lower()
   fmf_id = tmt_object.fmf_id

   url_params: dict[str, str] = {
     f'{tmt_type}-url': fmfid.url,
     f'{tmt_type}-name': fmfid.name
   }

   if fmf_id.path is not None:
     url_params[f'{tmt_type}-path'] = fmfid.path
   if fmf_id.ref is not None:
     url_params[f'{tmt_type}-ref'] = fmfid.ref

   return url_params
   ```
3. `construct_url_from_list(url: str, url_parts: list[str])` takes `url` parameter which is unexpected - I'd suggest calling this one `baseurl`, it's a fairly common approach when you need to construct a URL from some starting website URL plus a path and/or parameters.
4. `joined_url += '&format=html'` - can we move this one to the `create_url()` function? It seems to be yet another parameter, `url_params['format'] = 'html'`.
5. https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urlencode is exactly the helper we need to take parameters and convert them into a part of URL. It takes care of encoding weird characters, joining them with `&`, and all that stuff. When doing it manually, you always have to remember to put `&` at the right place, and you do not encode characters that need to be encoded, like spaces or `/`.
6. `joined_url = url + joined_url` - and if the service base URL does not end with `/`, this will produce invalid URL.
7. the same applies to *parsing* URL, see https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urlparse - much better choice than `split()` and indexing. For one call, you get a named tuple and can inspect schema, hostname, all the components, with no chance of getting an index wrong by mistake.

As you can see, it looks trivial, but there are a lot of traps, encoding being the biggest one. Luckily, Python provides a very helpful library. Eventually, you can end up with something much simpler and safer, e.g.

```python
def create_url_params(tmt_object: 'tmt.base.Core') -> dict[str, str]:
   tmt_type = tmt_object.__class__.__name__.lower()
   fmf_id = tmt_object.fmf_id

   url_params: dict[str, str] = {
     'format': 'html',
     f'{tmt_type}-url': fmfid.url,
     f'{tmt_type}-name': fmfid.name
   }

   if fmf_id.path is not None:
     url_params[f'{tmt_type}-path'] = fmfid.path

   if fmf_id.ref is not None:
     url_params[f'{tmt_type}-ref'] = fmfid.ref

   return url_params

def create_url(baseurl: str, url_params: dict[str, str]) -> str:
    return urllib.parse.urljoin(baseurl, urllib.parse.urlencode(url_params))
```

I'm not sure what's the purpose of `service_url` as a list, it seems to collect URL parameters for various nodes? `if len(nodes) > 1 and not separate:`, `url_part = create_url(tmt_object=node)` -> `service_url.extend(url_part)`, in the next iteration, the list is empty and will get yet another string from new call to `create_url(tmt_object=node)`?